File: src/main/java/me/markoutte/sandbox/swing/monsters/MonsterMark.java
package me.markoutte.sandbox.swing.monsters;

import me.markoutte.sandbox.swing.SwingExample;
import me.markoutte.sandbox.swing.SwingPerformance;

import javax.imageio.ImageIO;
import javax.swing.*;
import javax.validation.constraints.NotNull;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;

public class MonsterMark {

    private static final Image[] SPRITES = Arrays.stream(Monsters.values())
            .map(monsters -> SpriteImage32.createOne(monsters.getBase64(), bytes -> {
                try (ByteArrayInputStream bais = new ByteArrayInputStream(bytes)) {
                    return ImageIO.read(bais);
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            })).toArray(Image[]::new);

    public static void main(String[] args) {
        SwingUtilities.invokeLater(() -> {
            final Painter.Policy policy = Painter.Policy.TIMER_DRIVEN;
            switch (policy) {
                //noinspection ConstantConditions
                case TIMER_DRIVEN:
                    SwingPerformance.windowsTimerHack();
                    break;
                // must be called before AWT is initialized
                // see javax.swing.BufferStrategyPaintManager
                case VSYNC_DRIVEN:
//                System.setProperty("sun.java2d.opengl", "true");
                    System.setProperty("swing.bufferPerWindow", "true");
                    break;
            }

            Painter painter = new Painter(policy);
            painter.addMouseListener(new MouseAdapter() {
                @Override
                public void mouseClicked(MouseEvent e) {
                    painter.requestNew = 1000;
                }
            });
            JFrame w = SwingExample.inFrame("Monsters Swing", painter);
            w.setVisible(true);
            painter.requestNew = 1;

            //noinspection ConstantConditions
            if (policy == Painter.Policy.VSYNC_DRIVEN) {
                SwingPerformance.enableVsyncHack(w);
            }
        });
    }

    private static class Painter extends JComponent {
        private final EventCounter fps = new EventCounter(10000);
        private final List<SpriteImage32<Image>> cages = new ArrayList<>();
        private double counter = 0;
        private int requestNew = 0;
        private long time = System.nanoTime();
        private final long nanos = TimeUnit.SECONDS.toNanos(1);
        @SuppressWarnings("FieldCanBeLocal")
        private final int px = 50; // pixels per second
        private final JLabel info = new JLabel();
        private final @NotNull Policy updatePolicy;

        enum Policy { TIMER_DRIVEN, VSYNC_DRIVEN }

        public Painter(Policy updatePolicy) {
            this.updatePolicy = updatePolicy;
            if (updatePolicy == Policy.TIMER_DRIVEN) {
                Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(this::schedule, 0, 500, TimeUnit.MICROSECONDS);
            }

            setOpaque(true);
            setBackground(Color.WHITE);
            setLayout(null);
            add(info);

            info.setBorder(BorderFactory.createEmptyBorder(5, 5, 5, 5));
            info.setOpaque(true);
            info.setBackground(Color.WHITE);
        }

        private void schedule() {
            long current = System.nanoTime();
            counter += (current - time) / (double) nanos * px;
            if (counter > px * 10_000) {
                counter %= px * 10_000;
            }
            time = current;
            info.setText(String.format("%d monsters, fps = %d", cages.size(), fps.size()));
            info.setBounds(5, 5, info.getPreferredSize().width, info.getPreferredSize().height);

            repaint();
        }

        @Override
        protected void paintComponent(Graphics g) {
            Graphics2D g2d = (Graphics2D) g.create();
            Rectangle bounds = g2d.getClipBounds();
            g2d.setColor(getBackground());
            g2d.fillRect(0, 0, bounds.width, bounds.height);

            checkRequestAndAddIfNeeded();

            long s = Math.round(counter);
            for (SpriteImage32<Image> cage : cages) {
                Point p = cage.calcPoint(s, bounds.width, bounds.height, 1f);
                long frame = time % nanos / (nanos / cage.source.getWidth(null) * SpriteImage32.SIZE);
                int offset = (int) (frame * SpriteImage32.SIZE);
                g2d.drawImage(cage.source,
                        p.x, p.y, p.x + cage.width, p.y + cage.height,
                        offset, 0, offset + cage.width, cage.height,
                        null
                );
            }

            g2d.dispose();
            fps.update();

            if (updatePolicy == Policy.VSYNC_DRIVEN) {
                schedule();
            }
        }

        private static final Random R = new Random();

        private void checkRequestAndAddIfNeeded() {
            if (requestNew > 0) {
                while (requestNew-- > 0) {
                    cages.add(SpriteImage32.catchOne(SPRITES[R.nextInt(SPRITES.length)]));
                }
            }
        }
    }
}

File: src/main/java/me/markoutte/sandbox/swing/monsters/MonsterMark.kt
package me.markoutte.sandbox.swing.monsters

import androidx.compose.desktop.Window
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.material.Text
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.currentRecomposeScope
import androidx.compose.runtime.remember
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.drawIntoCanvas
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import org.jetbrains.skija.Image
import org.jetbrains.skija.Rect
import java.util.concurrent.TimeUnit
import kotlin.math.roundToLong
import kotlin.random.Random

fun main() = Window(title = "Monsters Compose", size = IntSize(640, 480)) {

    val state = remember { GameState() }
    val scope = currentRecomposeScope

//   https://github.com/JetBrains/compose-jb/issues/153#issuecomment-817831381
   LaunchedEffect(state) {
       while (true) withFrameNanos {
           state.update(it)
           scope.invalidate()
       }
   }

    Canvas(modifier = Modifier
        .fillMaxSize()
        .background(Color.White)
        .onSizeChanged {
            state.size = IntSize(it.width, it.height)
        }
        .clickable(
            onClick = {
                repeat(1000) { cages.add(catchOne()) }
                state.count = cages.size
            },
            // don't draw an animation on click
            indication = null,
            interactionSource = remember { androidx.compose.foundation.interaction.MutableInteractionSource() })
    ) {
        val s = (state.counter * density).roundToLong()
        cages.forEach { cage ->
            val p = cage.calcPoint(s, state.size.width, state.size.height, density)
            val frame = state.time % state.nanos / (state.nanos / cage.source.width * SpriteImage32.SIZE)
            val offset = frame * SpriteImage32.SIZE
            drawIntoCanvas {
                it.nativeCanvas.drawImageRect(
                    cage.source,
                    Rect.makeXYWH(
                        offset.toFloat(),
                        0f,
                        cage.width.toFloat(),
                        cage.height.toFloat()
                    ),
                    Rect.makeXYWH(
                        p.x.toFloat(),
                        p.y.toFloat(),
                        cage.width.toFloat() * density,
                        cage.height.toFloat() * density
                    ),
                    null,
                    true
                )
            }
        }
        state.fps = repaints.update()
    }

    Text("${state.count} monsters, fps = ${state.fps}",
        modifier = Modifier
            .offset(5.dp, 5.dp)
            .background(Color.White)
            .padding(5.dp)
    )
}

private val repaints = EventCounter(10000)

// we don't use mutable states here, because scope will be invalidate manually
private class GameState {
    var size = IntSize(200, 200)
    var count = cages.size
    var fps = 0
    var counter = 0.0
        private set
    var time = System.nanoTime()
        private set
    val px = 50 // pixels per second
    val nanos = TimeUnit.SECONDS.toNanos(1)

    fun update(current: Long) {
        counter += (current - time) / nanos.toDouble() * px
        if (counter > px * 10000) {
            counter %= (px * 10000).toDouble()
        }
        time = current
    }
}

private val sprite = Monsters.values().map {
    SpriteImage32.createOne(it.base64, Image::makeFromEncoded)
}

private fun catchOne() = SpriteImage32.catchOne(sprite[Random.nextInt(sprite.size)])

private val cages = mutableListOf(catchOne())